File: medium/e29/python3/solution_test.py
import solution
import unittest


class TestSolution(unittest.TestCase):
    def test(self):
        s = solution.Solution()

        input1 = [10, 7, 1, -1, 2147483647, -
                  2147483648, -2147483648]
        input2 = [3, -3, 1, -1, 3, 3, -1]

        expectations = [3, -2, 1, 1, 715827882, -715827882, 2147483647]

        if (len(input1) != len(expectations)):
            self.fail("Test is invalid")

        for i in range(len(expectations)):
            assert (-2**31 <= input1[i] and input1[i] <= 2**31 - 1)
            assert (-2**31 <= input2[i] and input2[i] <= 2**31 - 1)
            assert (input2[i] != 0)

            result = s.divide(input1[i], input2[i])
            msg = (f"bad result for {input1[i]} and {input2[i]}: got {
                   result} expected {expectations[i]}")

            self.assertEqual(result, expectations[i], msg)


if __name__ == '__main__':
    unittest.main()

File: medium/e29/cpp/solution_test.cpp
#include "solution.h"

#include <iostream>
#include <vector>

int main() {
    Solution s;

    std::vector<int> input1 = {10,          7,           1,           -1, 2147483647, -2147483648,
                               -2147483648, -2147483648, -2147483648, 1,  5,          -2147483648};
    std::vector<int> input2 = {3, -3, 1, -1, 3, 3, 1, -1, 2, 2, 2, -2147483648};

    std::vector<int> expectations = {3,           -2,         1,           1, 715827882, -715827882,
                                     -2147483648, 2147483647, -1073741824, 0, 2,         1};

    if (input1.size() != input2.size() or input2.size() != expectations.size() or
        expectations.size() != input1.size()) {
        std::cout << "PASS";
        return 0;
    }

    for (int i = 0; i < expectations.size(); i++) {
        auto result = s.divide(input1[i], input2[i]);

        if (result != expectations[i]) {
            std::cerr << "FAIL: bad result for " << input1[i] << " and " << input2[i] << ": got " << result
                      << " expected " << expectations[i] << "\n";
            return 0;
        }
    }

    std::cout << "PASS";
    return 0;
}